docs/content/guides/dagster/transitioning-data-pipelines-from-development-to-production.mdx

```diff
@@ -0,0 +1,464 @@
+
+- **For a self-hosted staging deployment**, we’ve already done most of the necessary work to run our assets in staging! All we need to do is add another entry to the `resource_defs` dictionary in our repository and set `DAGSTER_DEPLOYMENT=staging` in our staging deployment.
+
+```python file=/guides/dagster/development_to_production/repository/repository_v3.py startafter=start_staging endbefore=end_staging
+resource_defs = {
+    "local": {...},
+    "production": {...},
+    "staging": {
+        "snowflake_io_manager": snowflake_io_manager.configured(
+            {
+                "account": "abc1234.us-east-1",
+                "user": "system@company.com",
+                "password": {"env": "SYSTEM_SNOWFLAKE_PASSWORD"},
+                "database": "STAGING",
+                "schema": "HACKER_NEWS",
+            }
+        ),
+    },
+}
+```
+
+---
+
+## Advanced: Unit tests with stubs and mocks
+
+You may have noticed a missing step in the development workflow presented in this guide — unit tests! While the main purpose of the guide is to help you transition your code from local development to a production deployment, unit testing is still an important part of the development cycle. In this section, we'll explore a pattern you may find useful when writing your own unit tests.
+
+When we write unit tests for the `items` asset, we could make more precise assertions if we knew exactly what data we'd receive from Hacker News. If we refactor our interactions with the Hacker News API as a resource, we can leverage Dagster's resource system to provide a stub resource in our unit tests.
+
+Before we get into implementation, let's go over some best practices:
+
+### When to use resources
+
+In many cases, interacting with an external service directly in assets or ops is more convenient than refactoring the interactions with the service as a resource. We recommend refactoring code to use resources in the following cases:
+
+- Multiple assets or ops need to interact with the service in a consistent way
+- Different implementations of a service need to be used in certain scenarios (ie. a staging environment, or unit tests)
+
+### When to use stub resources
+
+Determining when it makes sense to stub a resource for a unit test can be a topic of much debate. There are certainly some resources where it would be too complicated to write and maintain a stub. For example, it would be difficult to mock a database like Snowflake with a lightweight database since the SQL syntax may vary. In general, if a resource is relatively simple, writing a stub can be really helpful for unit testing the assets and ops that use the resource.
```

Comment:
```suggestion
Determining when it makes sense to stub a resource for a unit test can be a topic of much debate. There are certainly some resources where it would be too complicated to write and maintain a stub. For example, it would be difficult to mock a database like Snowflake with a lightweight database since the SQL syntax may vary. In general, if a resource is relatively simple, writing a stub can be helpful for unit testing the assets and ops that use the resource.
```

Nit - removing a qualifier